src/python/torchdistx/slow_momentum/slow_momentum_optimizer.py

```diff
@@ -0,0 +1,222 @@
+        """
+        optim_state_dict = self._base_optim.state_dict()
+        optim_state_dict["slowmo_freq"] = self.slowmo_freq
+        optim_state_dict["slowmo_factor"] = self.slowmo_factor
+        optim_state_dict["slowmo_lr"] = self.slowmo_lr
+        optim_state_dict["step"] = self.averager.step
+
+        return optim_state_dict
+
+    def load_state_dict(self, state_dict):
+        r"""
+        This is the same as :class:`torch.optim.Optimizer`
+        :meth:`load_state_dict`, but also restores Slow Momentum's
+        specific parameters, saved in the provided ``state_dict``.
+        Additionally, it restors ``base_lr``, which refers to
+        the base optimizer's learning rate, and re-initializes slow momentum
+        buffers.
+        """
+        self.slowmo_freq = state_dict["slowmo_freq"]
+        self.averager.period = state_dict.pop("slowmo_freq")
+        self.slowmo_factor = state_dict.pop("slowmo_factor")
+        self.slowmo_lr = state_dict.pop("slowmo_lr")
+        self.averager.step = state_dict.pop("step")
+        self._base_optim.load_state_dict(state_dict)
+        # check that base optimizer's learning rate is stored in param_groups
+        if not (self.param_groups and self.param_groups[0]["lr"]):
+            raise ValueError(
+                "Base optimizer does not have parameters or learning rate specified."
+            )
+        self.base_lr = self.param_groups[0]["lr"]
+        self._init_slowmo_buffers()
+
+    def step(self):
+        r"""
+        Performs a single optimization step (parameter update)
+        and a slow momentum update. Slow momentum update involves
+        model's exact averaging of parameters and a momentum update,
+        which happens every `slowmo_freq` step.
+        """
+        self._base_optim.step()
+        # Averager averages parameters between workers every `slowmo_freq` step.
+        # At other times it just increases step counter.
+        self.averager.average_parameters(params=self.param_groups)
+        # Since at this poin averager has increased its step,
+        # we need to check (self.averager.step - 1).
+        # No need to do momentum step at step 0
+        if (self.averager.step - 1) % self.slowmo_freq == 0 and self.averager.step != 1:
+            for group in self.param_groups:
+                for param in group["params"]:
+                    # Update the slow momentum
+                    p_state = self.state[param]
+                    factor = 1 / self.base_lr
+
+                    p_state["slow_momentum"].mul_(self.slowmo_factor).sub_(
+                        param.data, alpha=factor
+                    ).add_(p_state["prev_param"], alpha=factor)
+
+                    # Update parameters
+                    p_state["prev_param"].add_(
+                        p_state["slow_momentum"], alpha=-self.slowmo_lr * self.base_lr
+                    )
+                    param.data.copy_(p_state["prev_param"])
```

Comment:
I think also no need for `.data` before `copy_()` with the same caveat about `torch.no_grad()` as above